ethcore/src/engines/tendermint/mod.rs

```diff
@@ -101,6 +101,53 @@ pub struct Tendermint {
 	validators: Box<ValidatorSet>,
 }
 
+struct EpochVerifier {
+	subchain_validators: SimpleList,
+}
+
+fn combine_proofs(signal_number: BlockNumber, set_proof: &[u8]) -> Vec<u8> {
+	let mut stream = ::rlp::RlpStream::new_list(3);
+	stream
+		.append(&signal_number)
+		.append(&set_proof)
+		.out()
+}
+
+fn destructure_proofs(combined: &[u8]) -> Result<(BlockNumber, &[u8]), Error> {
+	let rlp = UntrustedRlp::new(combined);
+	Ok((
+		rlp.at(0)?.as_val()?,
+		rlp.at(1)?.data()?,
+	))
+}
+
+impl super::EpochVerifier for EpochVerifier {
+	fn verify_light(&self, header: &Header) -> Result<(), Error> {
+		let mut signatures = HashSet::new();
+		let ref header_signatures_field = header.seal().get(2).ok_or(BlockError::InvalidSeal)?;
+		for rlp in UntrustedRlp::new(header_signatures_field).iter() {
+			let signature = rlp.as_val()?; // TODO: why is `rlp` have type ()?
+			signatures.insert(signature);
+		}
+
+		let n = signatures.len();
+		let threshold = self.subchain_validators.len() * 2/3;
+		if n > threshold {
+			Ok(())
+		} else {
+			Err(EngineError::BadSealFieldSize(OutOfBounds {
+				min: Some(threshold),
+				max: None,
+				found: n
+			}).into())
+		}
+	}
+
+	fn check_finality_proof(&self, proof: &[u8]) -> Option<Vec<H256>> {
+		panic!("Tendermint blocks have instant finality, so `check_finality_proof` should never get called.")
```

Comment:
tendermint blocks do have instant finality, so the "proof" here would just be an encoded header (which we check for seal validity).

the way warp sync for PoA works is by providing a list of validator set changes along with proofs of finality of the change. A change is signalled, and the old validator set approves the signal by finalizing the block where the signal occurred. From this proved list of validator set handoffs we can get to the most recent finalized block with full security. `check_finality_proof` is the way we check that signals were finalized, so it can't panic.